src/dmd/mars.d

```diff
@@ -869,20 +869,109 @@ private int tryMain(size_t argc, const(char)** argv)
 
     // inlineScan incrementally run semantic3 of each expanded functions.
     // So deps file generation should be moved after the inlinig stage.
-    if (global.params.moduleDeps)
+    // inlineScan incrementally run semantic3 of each expanded functions.
+    // So deps file generation should be moved after the inlinig stage.
+    if (global.params.moduleDeps || global.params.doMakeDeps)
     {
         foreach (i; 1 .. modules[0].aimports.dim)
             semantic3OnDependencies(modules[0].aimports[i]);
 
-        OutBuffer* ob = global.params.moduleDeps;
-        if (global.params.moduleDepsFile)
+        if (global.params.moduleDeps)
         {
-            auto deps = File(global.params.moduleDepsFile);
-            deps.setbuffer(cast(void*)ob.data, ob.offset);
-            writeFile(Loc.initial, &deps);
+            OutBuffer* ob = global.params.moduleDeps;
+            if (global.params.moduleDepsFile)
+            {
+                auto deps = File(global.params.moduleDepsFile);
+                deps.setbuffer(cast(void*)ob.data, ob.offset);
+                writeFile(Loc.initial, &deps);
+            }
+            else
+                printf("%.*s", cast(int)ob.offset, ob.data);
+        }
+        if (global.params.doMakeDeps)
+        {
+            // Recursive dependencies scanner, fills string tables with module source paths and string-imported files paths
+            void recursivelyAddModuleDependencies(Module m, StringTable* moduleImportedPaths, StringTable* stringImportedPaths)
+            {
+                const(char)* mpath = m.srcfile.toChars();
+                // Skip modules already in the table, add self
+                if(moduleImportedPaths.insert(mpath, strlen(mpath), null) is null)
```

Comment:
strlen must die.

Reply:
Agreed, but what's the alternative when you have a `char*`?

Reply:
I thought we already have a property that returns a D string in Identifier?

Reply:
No, not from `dmd.root.filename.FileName` (which is where `mpath` originates), as far as I can see.

Ideally `dmd.root.filename.FileName` should be updated to store the length as well and not only the pointer.

Reply:
If  you don't what to completely reimplement `FileName`, perhaps it's acceptable to add a `toString` method to `FileName` that encapsulates `strlen`. This way we at least doesn't increase the surface of where `char*` is used.

Reply:
I followed Jacob's suggestion with the `toString` method, unfortunately there still remains a use of `strlen` a couple of lines below, because it indexes into a `Strings` member of the `Module` class, which uses `const(char)*` as it's exposed in the C header file, please correct me if I'm wrong and I could change that into an array of D strings.